docs/architecture-decisions/2023-01-12_smart-contract-deployment.md

```diff
@@ -0,0 +1,375 @@
+
+There are techniques that can be used to allow for immutable smart contracts, but let them be evolved somewhat:
+
+- You could release a new version of a smart contract and include an ability for that smart contract to communicate to/from the existing smart contract to migrate state.
+- You could ensure that clients can dynamically find the latest smart contract so the application ID / address doesn't need to be hardcoded and the smart contract remains addressable.
+  - This could be done via some sort of on-chain or off-chain lookup, and/or by encoding information into the creation transaction note of the smart contract app.
+- You could limit MainNet releases to major upgrades that happen infrequently and let users opt-in to whether or not they use the new version or not, and having a set of calls the user can sign to migrate their state/value from one contract to the next.
+
+Lastly, it's worth noting that having fast development feedback loops during development and testing of smart contracts is likely a very useful feature to improve the development experience and speed. For this reason, allowing contracts to be upgraded or at the very least deleted (and then recreated) is likely very useful, but potentially when deployed to MainNet a switch could be made to disallow upgrading / deleting (as relevant).
+
+The goal of AlgoKit is to create a development experience that is productive and easy, but also one that is secure by default and helps developers fall into the pit of success. For that reason, and given the consequences of getting this wrong the suggested approach AlgoKit takes is:
+
+- All provided smart contract templates are by default immutable
+- An immutability automated test is included by default to ensure that smart contracts can't be upgraded by the contract creator (this would have to be deleted by a developer, who is then opting in to the consequences of that)
+- All provided smart contract templates are by default permanent when deployed to MainNet, but deletable elsewhere to facilitate an iterative development experience
+- Client code will include mechanisms to dynamically find deployed applications in LocalNet and TestNet environments to support delete/recreate flows and improve the developer experience
+- MainNet deployments will immediately (i.e. before any usage occurs) check that smart contracts are not upgradeable by the creator account by default (with an explicit opt-out option available for smart contracts that are meant to be upgradeable, which in turn will issue a warning to the developer to explain the implications)
+- MainNet deployments will immediately (i.e. before any usage occurs) check that smart contracts are not deletable by the creator account by default (with an explicit opt-out option available for smart contracts that are meant to be deletable, which in turn will issue a warning to the developer to explain the implications)
+
```

Comment:
+1 to the making contract immutable by default, so glad to see that. 

As for the various checks on that topic, what about checking that any inner app calls are calling apps that are immutable? Was that considered?

Reply:
Actually how do you check whether a contract if not upgradeable/deletable?
I'm guessing with beaker, this can be done relatively easily.
But in general, this may require formal analysis. I'm guessing that checking that inner app calls are to immutable apps can be done without formal analysis. This is maybe something that would require a static analysis tool.

Reply:
It's a great question @fabrice102!

I think initially we can:

a) ask the developer to specify if the contract is upgradeable/permanent
b) include an automated test that calls the contract with a basic check

Longer term, formal analysis would be ideal, but that's more advanced than would be possible for v1.